src/System.Private.CoreLib/shared/System/Threading/Thread.cs

```diff
@@ -210,6 +211,55 @@ public void Resume()
         public static void BeginThreadAffinity() { }
         public static void EndThreadAffinity() { }
 
+        // The upper bits of t_currentProcessorIdCache are the currentProcessorId. The lower bits of
+        // the t_currentProcessorIdCache are counting down to get it periodically refreshed.
+        // TODO: Consider flushing the currentProcessorIdCache on Wait operations or similar
+        // actions that are likely to result in changing the executing core
+        [ThreadStatic]
+        private static int t_currentProcessorIdCache;
+
+        private const int ProcessorIdCacheShift = 16;
+        private const int ProcessorIdCacheCountDownMask = (1 << ProcessorIdCacheShift) - 1;
+        private const int ProcessorIdRefreshRate = 5000;
+
+        private static int RefreshCurrentProcessorId()
+        {
+            int currentProcessorId = GetCurrentProcessorNumber();
+
+            // If GetCurrentProcessorNumber() is not fully implemented it will return -1.
+            // On Unix, GetCurrentProcessorNumber() is implemented in terms of sched_getcpu, which
+            // doesn't exist on all platforms.  On those it doesn't exist on, GetCurrentProcessorNumber()
+            // returns -1.  As a fallback in that case and to spread the threads across the buckets
+            // by default, we use the current managed thread ID as a proxy.
+            if (currentProcessorId < 0) currentProcessorId = Environment.CurrentManagedThreadId;
+
+            // Ensure the Id is in range of the ProcessorCount
+            currentProcessorId = (int)((uint)currentProcessorId % (uint)Environment.ProcessorCount);
```

Comment:
I believe in a VM proc Id could be above ProcessorCount. Do we care?

Reply:
Do you as to whether its accurate vs in range 0-ProcCount?

Reply:
Do you as to whether its accurate vs in range 0-ProcCount?

The first means a mod is needed for each use; the second a mod every 5000 calls.

Reply:
Another concern is - is this cheap? The number of cores could be pretty odd nowdays. 
I generally round up the number of stripes to the next ^2 - so that I could use binary masking which is much cheaper. 

In that case some slots may not be used due to the oddness of the proc number (or due to affinity), but that is never a big deal.

Reply:
Also the mod will be changed to not be a mod at Tier1 as the divisor will be a constant.

Reply:
> the second a mod every 5000 calls.

@VSadov has done a number of experiments around this and we believe that this limit should be <50.

On more recent hardware, none of this TLS caching should be needed and this should just call `RDPID` instruction that is both fastest and most accurate.

Reply:
On modern CPUs fetching the core id is very cheap. I think we will not cache in those case, since cache would be slower than actually getting the precise Id. 
Then, I am afraid, this `%` will show up as the most expensive part of the operation and the caller can't do anything about that.

Reply:
> Another concern is - is this cheap?

Becomes a multiply + shift at Tier1 (certainly below 64); and the callsites that are changing are currently using mod

Reply:
> Then, I am afraid, this `%` will show up as the most expensive part of the operation and the caller can't do anything about that.

However the Jit can because `Environment.ProcessorCount` is a runtime constant after https://github.com/dotnet/coreclr/pull/27543

For AoT could calculate the fastmod divisor at the same time as setting `ProcessorCount` and use `FastMod` as per Dictionary https://github.com/dotnet/coreclr/pull/27543

Reply:
Not sure how (if you can) you mark a path as the R2R/AoT version, however...

Reply:
I think guaranteeing `[0..ProcessorCount)`  range is unnecessary burden for this API. 
The original idea that this is just a number that correlates with most recent core is more flexible.

It should be up to callers to reduce this to the range that they need.
We do not know what that range is here - is it `ProcessorCount`,  `ProcessorCount/2`, `Min(ProcessorCount, 8)`,    `RoundToPow2(ProcessorNumber)` or whatever else... .

Reply:
> It should be up to callers to reduce this to the range that they need. We do not know what that range is here 

Nor does the user because the implementation changes per platform; it also currently adds 100 to it, so I'm not sure how a caller can sensibly reduce it to the range without paying attention to the implementations which then makes the range reduction leaky and fragile?

Reply:
If the exact coreid is important over an in range value it would be better to expose `RDPID` or similar via intrinsics to expose that? Then you can check the `.IsSupported` flags etc

Reply:
`CurrentProcessorId` is platform neutral concept. I do not think we need to have yet another set of hardware intrinsics for it.

> it also currently adds 100 to it

I would be ok with dropping this and making this API to return the underlying Processor ID.

Reply:
I'm happy to push the mod bypasses out into the callsites rather than in `GetCurrentProcessorId`

> `CurrentProcessorId` is platform neutral concept.

However, my point is it doesn't return a platform neutral value

Reply:
I think we could drop the `+100` part.
It is not a big problem for reducer though. Until you have 100+ cores in the system any method of reducing - modding, masking, hashing, should work the same. 
After 100+ cores, yes, it could be a bit of a problem for those who use masking.

Reply:
Just changed the call sites in https://github.com/dotnet/coreclr/pull/27600 if that is preferred?